cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java

```diff
@@ -231,35 +240,46 @@ public void starting() {
           // If RunId is not time-based, use current time as start time
           startTimeInSeconds = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
         }
-        if (workflowName == null) {
+        if (workflowRunId == null) {
           store.setStart(program.getId(), runId.getId(), startTimeInSeconds, twillRunId,
                          userArgs.asMap(), arguments.asMap());
         } else {
           // Program started by Workflow
-          store.setWorkflowProgramStart(program.getId(), runId.getId(), workflowName, workflowRunId, workflowNodeId,
-                                        startTimeInSeconds, twillRunId);
+          store.setWorkflowProgramStart(program.getId(), runId.getId(), workflowRunId.getProgram(),
+                                        workflowRunId.getRun(), workflowNodeId, startTimeInSeconds, twillRunId);
         }
       }
 
       @Override
       public void terminated(Service.State from) {
         closeAllQuietly(closeables);
+        ProgramRunStatus runStatus = ProgramController.State.COMPLETED.getRunStatus();
         if (from == Service.State.STOPPING) {
           // Service was killed
+          runStatus = ProgramController.State.KILLED.getRunStatus();
+        }
+
+        if (workflowRunId == null) {
           store.setStop(program.getId(), runId.getId(), TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()),
-                        ProgramController.State.KILLED.getRunStatus());
+                        runStatus);
         } else {
-          // Service completed by itself.
-          store.setStop(program.getId(), runId.getId(), TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()),
-                        ProgramController.State.COMPLETED.getRunStatus());
+          store.setWorkflowProgramStop(workflowRunId, programRunId,
+                                       TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()), runStatus, null);
         }
       }
 
       @Override
       public void failed(Service.State from, Throwable failure) {
         closeAllQuietly(closeables);
-        store.setStop(program.getId(), runId.getId(), TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()),
-                      ProgramController.State.ERROR.getRunStatus());
+        if (workflowRunId == null) {
```

Comment:
This is almost identical to terminate. Maybe better to have a private method in the listener to do so.